### PR TITLE
Adding Events support in bid responses

### PR DESCRIPTION
**Pull request by laurb9**
Implemented wurl insertion in cached bids, vast rewriting and bid.ext.prebid.events for #1015 
Updated and refactored auction_test unit tests for cached bids
Updated exchange_test for bid.ext.prebid.events and vast rewriting

Test coverage
>ok  	github.com/prebid/prebid-server/exchange	6.039s	coverage: 90.5% of statements
>ok  	github.com/prebid/prebid-server/endpoints/events	0.022s	coverage: 97.4% of statements

PG is deferred to later PR (PBS-Java does not implement it either)

CC @bretg

### CODE REVIEW

**Comment by laurb9** [comment]
Rebased after #1532

File: endpoints/events/vtrack.go
Comment (SyntaxNode) [inline endpoints/events/vtrack.go:279]: Why the change to an empty string from passing it through? aka.. why not continue returning `c`?

File: endpoints/events/vtrack.go
Comment (SyntaxNode) [inline endpoints/events/vtrack.go:274]: Nitpick: There's no need to repeat type declarations of arguments. This can be simplified to:
```
func ModifyVastXmlString(externalUrl, c, bidid, bidder, accountId string, timestamp int64) string {
```

File: endpoints/events/vtrack.go
Comment (SyntaxNode) [inline endpoints/events/vtrack.go:293]: Nitpick: Comment is outdated. `modifyVastXml` -> `ModifyVastXmlJSON`

File: endpoints/events/vtrack.go
Comment (SyntaxNode) [inline endpoints/events/vtrack.go:295]: `FIXME` Ah! Can you fix it before merging it in please?

File: exchange/events.go
Comment (SyntaxNode) [inline exchange/events.go:18]: Nitpick: Consider renaming to `eventTracking`?

File: exchange/events.go
Comment (SyntaxNode) [inline exchange/events.go:33]: What's with the comment here?

File: openrtb_ext/request.go
Comment (SyntaxNode) [inline openrtb_ext/request.go:26]: Can this be a strongly typed struct instead?

File: exchange/exchange.go
Comment (SyntaxNode) [inline exchange/exchange.go:55]: Do we really need the whole info, or perhaps just a boolean lookup for if vast is enabled for the bidders?

**Review by SyntaxNode — COMMENTED** [review]
I didn't review the tests in detail yet. Have some other comments first.

File: endpoints/events/vtrack.go
Comment (laurb9) [reply endpoints/events/vtrack.go:274]: Nice, didn't know that

File: endpoints/events/vtrack.go
Comment (laurb9) [reply endpoints/events/vtrack.go:279]: Empty string indicates no change was needed, which for `ModifyVastXmlJSON` means to skip the conversion back to RawMessage. I felt this was ok given that empty string is not a possible outcome from doing the replacement on any non-empty input.

File: exchange/events.go
Comment (laurb9) [reply exchange/events.go:18]: Will do. I had named it `eventsData` to parallel `targetingData` which has similar interface.

File: exchange/events.go
Comment (laurb9) [reply exchange/events.go:33]: This should be on the account line. It's a note I left regarding the algorithm making a mention to a channel event configuration for account, which I decided to leave out for now, to add in the next iteration.

> A new account-level configuration is proposed to define whether an account needs server-side events by channel. e.g account 111 gets AMP/App events, but not web events. Account 222 gets events for all 3 channels: Web/AMP/App.
If an account doesn't specify the channel config, the default is that AMP and App are "true" if events are enabled.

File: openrtb_ext/request.go
Comment (laurb9) [reply openrtb_ext/request.go:26]: It can, but it would be an empty struct, as the only possible value so far is `events: {}`.

File: exchange/exchange.go
Comment (laurb9) [reply exchange/exchange.go:55]: I thought to keep it generic, so if we add per-channel settings later, or need to look up other bidder settings, we would not need to go back and modify the `exchange` struct and all the unit tests.

Performance-wise, it looks equivalent, only one map lookup is done either way to reach the value.
Let me know your thoughts.

File: endpoints/events/vtrack.go
Comment (laurb9) [reply endpoints/events/vtrack.go:295]: I was debating fixing this as a case of doing it fast vs doing it right, since only the Go serializer insists on escaping html - but perhaps I should.

Comment (laurb9) [reply endpoints/events/vtrack.go:295]: vscode nags me about the naming of these functions. It says they should be `ModifyVastXMLJSON` and `ModifyVastXMLString` respectively (probably because it doesn't know VAST is also an acro). Thoughts ?

File: endpoints/events/vtrack.go
Comment (SyntaxNode) [reply endpoints/events/vtrack.go:279]: Gotcha. Do you think it would be clearer to add an "ok" boolean return, like Go does for map retrievals?

File: endpoints/events/vtrack.go
Comment (SyntaxNode) [reply endpoints/events/vtrack.go:295]: I don't have a strong opinion on the name. You're right. It should technically be called ModifyVASTXMLJSON.. but I mean... my eyes! I would just leave it as is and not worry.

Could you please fill me in on why escaping the HTML is a problem here? I just don't have the context. Anyway, you can customize the Go json encoder to turn this feature off:
https://play.golang.org/p/C79fK-_OKe5

File: exchange/events.go
Comment (SyntaxNode) [reply exchange/events.go:33]: Gotcha. Could you perhaps leave this out or better explain it? It looks like a mistake without that context.

File: openrtb_ext/request.go
Comment (SyntaxNode) [reply openrtb_ext/request.go:26]: Ah. Is this like caching where the presence is the trigger?

File: exchange/exchange.go
Comment (SyntaxNode) [reply exchange/exchange.go:55]: I wasn't worried about performance. Thought it might be simpler to only pass the information needed. I feel 50/50 on it. I imagine we'll refactor the bidderInfo map to some kind of bidderInfoService in the future anyway. I think we might need more context after I add in hardcoded aliases.

File: openrtb_ext/request.go
Comment (laurb9) [reply openrtb_ext/request.go:26]: `ext.prebid.cache` needs a value though (`vastxml` and/or `bids`). 
I expect that in time, so will `ext.prebid.events`, but for now it works as you said.

File: endpoints/events/vtrack.go
Comment (laurb9) [reply endpoints/events/vtrack.go]: Sorry, html escapes in Go are a red herring here and probably worth a separate discussion.
Regarding the issue at hand, PBS has no control over how adm was encoded, so it has to be able to handle it, so I replaced `string()` with a proper `UnMarshal()`

**Comment by laurb9** [comment]
I implemented part of the spec change mentioned here https://github.com/prebid/prebid-server/issues/1015#issuecomment-740138755 - the modified VAST creative should be returned in bid response, not only cache.

I moved the VAST rewriting from auction/doCache to exchange/HoldAuction with the associated unit test migration, as they use two different sets of json specs. There are still some unanswered questions regarding the requirement that targeting be enabled for VAST rewriting because only the winning bids are modified, or even if we should modify all the bids instead.

For now, this implementation requires targeting, until the requirements are clarified. This should be ok if most requests use GAM or some other network that needs the targeting keys anyway, so targeting would be enabled.

**Comment by laurb9** [comment]
@bretg Added 6797b3a (unit tests) and 5f3018a (implementation) for updated functionality, sans PG or channels, which can be added in a later PR.

Updates:
- all VIDEO bids, cached and returned, have modified VAST if account and bidder enabled - not just winning cached bids.
- all NON-VIDEO returned bids have `ext.prebid.events` if account or request enabled
- all cached NON-VIDEO bids have `wurl` patched in if account or request enabled.

ok  	github.com/prebid/prebid-server/exchange	6.039s	coverage: 90.5% of statements

**Comment by laurb9** [comment]
I'm going to hold off resolving conflicts from #1626 until all the other module renames I see as pending got in, or this PR gets the go-ahead, whichever comes first :)

**Comment by laurb9** [comment]
Updated to new spec (ext.prebid.events now overrides server config for both vast and banner)
Resolved conflicts merging the latest master branch.

File: exchange/events_test.go
Comment (bsardo) [inline exchange/events_test.go:10]: You've put together some nice integration tests to obtain coverage of `modifyBidsForEvents` and helper methods `isModifyingVASTXMLAllowed` and `modifyBidVAST`. Thoughts on whether we should add a unit test for `modifyBidsForEvents`? I imagine you figured it would be a pain to test via UT because of the complexity around creating the seat bid function inputs.

Comment (laurb9) [reply exchange/events_test.go:10]: I've left it out because its behavior and parameters changed with the spec changes and I started leaning on the exchangeJSON tests to validate them instead.
Since those cover the code anyway, I feel the specialized UT would add less value now.

File: exchange/exchangetest/events-vast-account-off-request-off.json
Comment (laurb9) [reply exchange/exchangetest/events-vast-account-off-request-off.json:141]: Correct. The adm would be generated only to modify it, so if we do not need to modify then it is left alone.
This felt a little inconsistent at first (if it's ok to generate sometimes, why not always do it) but I've made peace with it.

File: exchange/exchangetest/events-bid-account-off-request-off.json
Comment (laurb9) [reply exchange/exchangetest/events-bid-account-off-request-off.json:98]: These seem to have been added in 4fac314 and are indeed not used. Will remove.

File: endpoints/events/vtrack.go
Comment (laurb9) [reply endpoints/events/vtrack.go:302]: Thought we can just return the original `data` unmodified as it's already json.RawMessage.

File: exchange/eventscachetest/json-account-off-request-on.json
Comment (laurb9) [reply exchange/eventscachetest/json-account-off-request-on.json:29]: Works either way because they're the same in json, but I will make it `&` so it is consistent.

**Review by bsardo — APPROVED** [review]
This is looking good. I'll get a review from another team member this week.

File: exchange/auction_test.go
Comment (laurb9) [reply exchange/auction_test.go:212]: Better to rename ACCOUNT_ID to TEST_ACC_ID to avoid confusion.

File: endpoints/events/vtrack.go
Comment (laurb9) [reply endpoints/events/vtrack.go:300]: Looks like it could work;  I must have written it originally as a single `if` where the other one was shadowed, and it ended like this

File: exchange/events.go
Comment (laurb9) [reply exchange/events.go:34]: This was meant for integration support, then I found out it is not yet implemented #1428 
I can remove it, but I would prefer to keep it to avoid integration popping up as a parameter later when it could be here. Thoughts ?

File: exchange/events.go
Comment (laurb9) [reply exchange/events.go:82]: It does not appear to be. It also means cached bids look different with respect to events than returned bids (cached bids have `wurl`, returned bids have`ext.prebid.events`).

File: exchange/events.go
Comment (laurb9) [reply exchange/events.go:88]: I thought it shouldn't happen because jsonBytes is valid json (it was marshalled earlier).
But I agree that ignored errors can waste time later, so I can add it to the error list.

File: exchange/events.go
Comment (laurb9) [reply exchange/events.go:93]: `&&` takes precedence and the expression is correct. I suppose I can also write it like this, if it's more readable:
```go
if !(ev.enabledForAccount || ev.enabledForRequest) || pbsBid.bidType == openrtb_ext.BidTypeVideo {
```

**Comment by laurb9** [comment]
Addressed comments and resolved conflicts from #1644

File: exchange/events.go
Comment (SyntaxNode) [reply exchange/events.go]: Ah ok. Gotcha. I think that's fine.

File: exchange/events.go
Comment (SyntaxNode) [reply exchange/events.go]: Thanks. The extra parenthesis makes it much easier to follow, IMHO.

**Review by SyntaxNode — APPROVED** [review]

**Review by bsardo — APPROVED** [review]
LGTM

**Pull request merged**
